lock_api/src/mutex.rs

```diff
@@ -508,13 +508,15 @@ impl<'a, R: RawMutex + 'a, T: ?Sized + 'a> MutexGuard<'a, R, T> {
     /// used as `MutexGuard::map(...)`. A method would interfere with methods of
     /// the same name on the contents of the locked data.
     #[inline]
-    pub fn map<U: ?Sized, F>(s: Self, f: F) -> MappedMutexGuard<'a, R, U>
+    pub fn map<U, F>(s: Self, f: F) -> MappedMutexGuard<'a, R, U>
     where
-        F: FnOnce(&mut T) -> &mut U,
+        F: FnOnce(&'a mut T) -> U,
```

Comment:
This is almost certainly wrong: `'a` is the lifetime of the `Mutex` object, which can be `'static`.
```suggestion
        F: FnOnce(&mut T) -> U,
```

Reply:
`'a` is the lifetime of the `MutexGuard`'s lock of the `Mutex`, which isn't the same and to my knowledge it is valid to get an `&'a mut T` from a `MutexGuard<'a, _, T>` since for the duration of `'a` the only reachable path to the `T` is via the `MutexGuard`, and the `MutexGuard` is consumed in this method. Can you come up with an example of a piece of invalid code that would be enabled by this?

Reply:
So I gave it some thought and I think I'll remove the lifetime, it's the more-conservative and more-obviously-correct option and it still allows for 99% of valid usecases, most importantly `MutexGuard::map(foo, |a| &mut a.some_field)`.

Reply:
```rust
    #[test]
    fn test_map() {
        let mut outer = &mut 0;
        static M: Mutex<i32> = Mutex::const_new(super::RawMutex::INIT, 0);
        let guard = super::MutexGuard::map(M.lock(), |inner: &'static mut i32| {
            outer = inner;
        });
        drop(guard);
        *outer = 1;
    }
```

Reply:
`'a` is the lifetime of the `Mutex`. The lifetime of the lock is the lifetime of `self`.

Reply:
Hmm after thinking about it some more, I think the change proposed in this PR can't possibly work with Rust's current lifetime system. Supporting this would require some form of support for higher-kinded types in the language.

Reply:
How do you mean? I'm not sure I understand what the issue is once you disallow `'a` access to the data.

Reply:
Well if you make the change that I suggested of removing the `'a` from the `FnOnce` then this code stops working:
```rust
    #[test]
    fn test_map() {
        let m = Mutex::new((0, 0));
        let guard = super::MutexGuard::map(m.lock(), |inner| &mut inner.0);
    }
```

```
error: lifetime may not live long enough
   --> src/mutex.rs:317:62
    |
317 |         let guard = super::MutexGuard::map(m.lock(), |inner| &mut inner.0);
    |                                                       ------ ^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                                       |    |
    |                                                       |    return type of closure is &'2 mut i32
    |                                                       has type `&'1 mut (i32, i32)`
```

Reply:
So I thought this could be fixed by adding a `'a` bound to `F`, but that doesn't work either:

```rust
#[test]
fn test_map() {
    use crate::{lock_api::RawMutex, Mutex, MutexGuard};

    const FOO: usize = 0;

    static OUTER: Mutex<&usize> = Mutex::const_new(RawMutex::INIT, &FOO);
    static M: Mutex<usize> = Mutex::const_new(RawMutex::INIT, 0);

    let guard = MutexGuard::map(M.lock(), |inner: &'static mut usize| {
        *OUTER.lock() = inner;
    });
    drop(guard);

    let outer: &usize = &*OUTER.lock();

    assert_eq!(*outer, 0);

    *M.lock() = 1;

    assert_eq!(*outer, 1);
}
```

Reply:
The issue is that we want the `U` that's returned from the function to be dependent on the input lifetime - we want to give the function a `&'a mut T` so that it's possible for it to return a field, but we _also_ don't want to let it rely on getting an `&'a mut T`. For references that's trivial - you can do `for<'any> F: FnOnce(&'any mut T) -> &'any mut U`, and in fact that's what the current code does - but for generic types that's not possible.

Reply:
Another issue is that it's not possible to restrict a type's lifetime generically - you can do `for<'any> F: FnOnce(&'any mut T) -> &'any mut U`, and then essentially convert your `&'a mut T` to a `for<'any> &'any mut T`, but you can't do the same if the `&'a mut T` is "hidden" by being some generic parameter. There's no equivalent of `for<'any> F: FnOnce(&'any mut T) -> &'any mut U` for `F: FnOnce(T) -> U`.

Reply:
This approach appears to work and disallows the code we want to disallow while allowing the code we want to allow, but the ergonomics are quite bad (I'll explain what the problem is below):

```rust
pub trait FnOnceShim<'a, T: 'a> {
    type Output: 'a;

    fn call(self, input: T) -> Self::Output;
}

impl<'a, F, In, Out> FnOnceShim<'a, In> for F
where
    F: FnOnce(In) -> Out,
    In: 'a,
    Out: 'a,
{
    type Output = Out;

    fn call(self, input: In) -> Self::Output {
        self(input)
    }
}

impl<'a, R: RawMutex + 'a, T: ?Sized + 'a> MutexGuard<'a, R, T> {
    // ...

    pub fn map<F>(
        s: Self,
        f: F,
    ) -> MappedMutexGuard<'a, R, <F as FnOnceShim<'a, &'a mut T>>::Output>
    where
        for<'any> F: FnOnceShim<'any, &'any mut T>,
    {
        let raw = &s.mutex.raw;
        let data = unsafe { &mut *s.mutex.data.get() };
        mem::forget(s);

        let data = f.call(data);
        MappedMutexGuard {
            raw,
            data,
            marker: PhantomData,
        }
    }
    
    // ...
}

impl<'a, R: RawMutex + 'a, T: 'a> MappedMutexGuard<'a, R, T> {
    // ...
    
    pub fn map<F>(s: Self, f: F) -> MappedMutexGuard<'a, R, <F as FnOnceShim<'a, T>>::Output>
    where
        for<'any> F: FnOnceShim<'any, T>,
    {
        let (data, raw) = {
            let s = mem::ManuallyDrop::new(s);
            (unsafe { ptr::read(&s.data) }, s.raw)
        };

        // `panic::catch_unwind` isn't available in `core`, so we use a dummy guard to unlock
        // the mutex in case of unwind.
        let lock_guard: MappedMutexGuard<'a, R, ()> = MappedMutexGuard {
            raw,
            data: (),
            marker: PhantomData,
        };

        let data = f.call(data);

        mem::forget(lock_guard);

        MappedMutexGuard {
            raw,
            data,
            marker: PhantomData,
        }
    }

    // ...
}
```

The issue is that Rust will no longer automatically convert the closure to the most-general version of it like it will when being passed as an `F: FnOnce(..) -> ..`. Therefore, you need to do something like this:

```rust
MutexGuard::map(M.lock(), { |inner| inner } as fn(&mut usize) -> &mut usize)
```

Reply:
Would it help if you changed the bounds to contain both `FnOnce` and `FnOnceShim`? Even though it is redundant, it may help with type inference.